test/object-store/sync/cloud_sync.cpp

```diff
@@ -0,0 +1,855 @@
+            realm::Property("dogs", PropertyType::Object | PropertyType::Array, "Dog"),
+            realm::Property("firstName", PropertyType::String), realm::Property("lastName", PropertyType::String),
+            realm::Property("realm_id", PropertyType::String | PropertyType::Nullable)});
+        config.schema = realm::Schema({AllTypesSyncObject::schema(), dog_schema, person_schema});
+        return config;
+    }
+
+    void run(std::function<void(SharedRealm)> block) {
+        auto realm = set_up();
+        block(realm);
+        tear_down(realm);
+    }
+private:
+    SharedRealm set_up() {
+        auto app = get_app_and_login(sync_manager.app());
+        auto config = setup_and_get_config(app);
+        opt_set_up(config);
+
+        auto realm = realm::Realm::get_shared_realm(config);
+        return realm;
+    }
+
+    void tear_down(SharedRealm) {
+        util::try_remove_dir_recursive(base_path);
+        util::try_make_dir(base_path);
+    }
+};
+
+template <PropertyType PT>
+void test_round_trip() {
+    static_assert(Corpus<PT>::value, "PropertyType must be added to corpus before testing");
+    // Add a new `AllTypesSyncObject` to the Realm
+    Harness().run([](auto realm) {
+        wait_for_download(*realm);
+        auto results = realm::Results(realm, realm->read_group().get_table("class_AllTypesSyncObject"));
+        realm->begin_transaction();
+        results.clear();
+        realm->commit_transaction();
+        wait_for_upload(*realm);
+
+        REQUIRE(results.size() == 0);
+        realm->begin_transaction();
+        const auto obj = AllTypesSyncObject(realm);
+        realm->commit_transaction();
+        REQUIRE(results.size() == 1);
+        wait_for_upload(*realm);
+    });
+    Harness().run([&](auto realm) {
+        wait_for_download(*realm);
+        // assert realm has at least one `AllTypesSyncObject`
+        auto results = realm::Results(realm, realm->read_group().get_table("class_AllTypesSyncObject"));
+        REQUIRE(results.size() == 1);
+
+        realm->begin_transaction();
+        auto obj = AllTypesSyncObject(results.get(0));
+        auto corpus = Corpus<PT>();
+        // assert the stored value is equal to the expected default value
+        REQUIRE(corpus.get(obj) == corpus.default_value);
+        corpus.set(obj, corpus.new_value);
+
+        // commit the changes
+        realm->commit_transaction();
+        wait_for_upload(*realm);
+    });
+    Harness().run([](auto realm) {
+        wait_for_download(*realm);
+        auto results = realm::Results(realm, realm->read_group().get_table("class_AllTypesSyncObject"));
+        REQUIRE(results.size() == 1);
+        auto obj = AllTypesSyncObject(results.get(0));
+        auto corpus = Corpus<PT>();
+        REQUIRE(corpus.get(obj) == corpus.new_value);
+    });
+}
+
+TEST_CASE("canonical_sync_corpus", "[sync][app]") {
+    // MARK: Int Round Trip
+    SECTION("Int Round Trip") {
```

Comment:
Can you use a [DYNAMIC_SECTION](https://github.com/catchorg/Catch2/blob/devel/docs/other-macros.md#test-case-related-macros) and pass in each of the corpus_types in a loop?

Reply:
Oh that's cool. Yeah, looks like that should work for every type except the Mixed types, but I can probably create a separate corpus_types for the mixed tests and do the same thing.